File: Scraping/copart_scraping.py
import os
import json
import time
import re
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
from selenium.common.exceptions import NoSuchElementException, ElementClickInterceptedException

# Set up Selenium options
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--no-sandbox")
options.add_argument("--disable-dev-shm-usage")

# Initialize WebDriver
service = Service(ChromeDriverManager().install())
driver = webdriver.Chrome(service=service, options=options)

url = "https://www.copart.com/lotSearchResults?free=true&query=&qId=9d03a1d2-1d10-4779-a6ab-497962109e20-1741568083803&emptySearch=true&index=undefined&searchCriteria=%7B%22query%22:%5B%22*%22%5D,%22filter%22:%7B%22MAKE%22:%5B%22lot_make_desc:%5C%22AUDI%5C%22%22%5D,%22YEAR%22:%5B%22lot_year:%5B2020%20TO%202025%5D%22%5D%7D,%22searchName%22:%22%22,%22watchListOnly%22:false,%22freeFormSearch%22:false%7D"

# Open the URL
driver.get(url)
time.sleep(10)

# Extract total search results
search_results_element = driver.find_element("css selector", "h3.text-black span.blue-heading")
total_results = int(search_results_element.text.replace(',', ''))
print(f"Total search results: {total_results}")

last_page_number = (total_results // 100) + (1 if total_results % 100 > 0 else 0)
print(f"Calculated last page number: {last_page_number}")

# Click the dropdown to open options
dropdown_trigger = driver.find_element("css selector", ".p-paginator-rpp-options")
dropdown_trigger.click()
time.sleep(2)

# Select "100" from the dropdown
option_100 = driver.find_element("css selector", "li[aria-label='100']")
option_100.click()
time.sleep(5)

def scrape_page(page_number):
    """Extracts car details from the current page and saves to a file."""
    cars = driver.find_elements("css selector", "tr.p-element.p-selectable-row.ng-star-inserted")
    print(f"Total cars found on page {page_number}: {len(cars)}")

    car_data = []

    for car in cars:
        try:
            # Extract Year, Make, Model
            title_element = car.find_element("css selector", "span.search_result_lot_detail")
            title = title_element.text.strip()
            title_parts = title.split(" ")

            year = title_parts[0]
            make = title_parts[1]
            model = title_parts[2]  # First word after Make

            # Extract link & image
            link_element = car.find_element("css selector", "a[aria-label='Lot Details']")
            car_link = f"https://www.copart.com{link_element.get_attribute('href')}"

            img_element = car.find_element("css selector", "a[aria-label='Lot Details'] img")
            main_image_url = img_element.get_attribute("src")

            # Extract lot number
            lot_number_element = car.find_element("css selector", ".search_result_lot_number")
            lot_number = lot_number_element.text.strip()

            # Extract Current Bid
            try:
                bid_container = car.find_element("css selector", "span.search_result_amount_block.text-black.p-bold")
                amount_element = bid_container.find_element("css selector", "span.currencyAmount")
                currency_element = bid_container.find_element("css selector", "span.currencyCode")

                amount_text = amount_element.text.strip()
                amount_value = re.sub(r'[^\d.]', '', amount_text)  # Extract only numeric value
                amount = float(amount_value) if amount_value else None
                currency = currency_element.text.strip()

                current_bid = {"Currency": currency, "Amount": amount}
            except NoSuchElementException:
                current_bid = None

            # Store data
            car_info = {
                "Title": title,
                "Make": make,
                "Year": year,
                "Model": model,
                "Link": car_link,
                "Main Image URL": main_image_url,
                "Lot Number": lot_number,
                "Current Bid": current_bid
            }
            print(car_info)
            car_data.append(car_info)

        except NoSuchElementException:
            print("Some elements were not found for this entry.")

    # Save the scraped data to a JSON file for this page
    filename = f'car_data_page_{page_number}.json'
    with open(filename, 'w') as f:
        json.dump(car_data, f, indent=4)
    print(f"Scraped data saved to '{filename}'")

# Scrape the first page
scrape_page(1)

# Iterate through all pages
current_page = 1
while current_page < last_page_number:
    try:
        next_button = driver.find_element("css selector", "button.p-paginator-next")
        if "p-disabled" in next_button.get_attribute("class"):
            print("Reached the last page.")
            break
        next_button.click()
        time.sleep(5)  # Wait for new page to load

        current_page += 1
        print(f"Scraping page {current_page}...")
        scrape_page(current_page)
    except (NoSuchElementException, ElementClickInterceptedException):
        print("Could not find or click the Next Page button.")
        break

# Close the browser
driver.quit()

### MERGE ALL JSON FILES ###
def merge_json_files():
    all_data = []
    page_number = 1

    while True:
        filename = f'car_data_page_{page_number}.json'
        if os.path.exists(filename):
            with open(filename, 'r') as f:
                data = json.load(f)
                all_data.extend(data)
            print(f"Merged data from {filename}")
            page_number += 1
        else:
            break  # Stop when no more files exist

    # Save the merged data into a single JSON file
    merged_filename = 'car_data_all.json'
    with open(merged_filename, 'w') as f:
        json.dump(all_data, f, indent=4)
    print(f"✅ All data merged into '{merged_filename}'")

# Run the merging function
merge_json_files()

File: Scraping/copart_scraping.js
const fs = require("fs");
const path = require("path");
const puppeteer = require("puppeteer");
const { v4: uuidv4 } = require("uuid");

// const StealthPlugin = require("puppeteer-extra-plugin-stealth");
// puppeteer.use(StealthPlugin());

const makes = [
    "AUDI",
    "BMW",
    "FERRARI",
    "LEXUS",
    "BENTLEY",
    "LAND ROVER",
    "ROLLS-ROYCE",
];

const yesterday = new Date();
yesterday.setDate(yesterday.getDate() - 1);

const today = new Date().toISOString().split("T")[0]; // YYYY-MM-DD
const formattedYesterday = yesterday.toISOString().split("T")[0];

const baseUrl =
    "https://www.copart.com/lotSearchResults?free=true&query=&searchCriteria=%7B%22query%22:%5B%22*%22%5D,%22filter%22:%7B%22MAKE%22:%5B%22lot_make_desc:%5C%22{make}%5C%22%22%5D,%22YEAR%22:%5B%22lot_year:%5B2020%20TO%202025%5D%22%5D%7D,%22watchListOnly%22:false,%22freeFormSearch%22:false%7D";

(async () => {
    const browser = await puppeteer.launch({ headless: true });
    const page = await browser.newPage();
    await page.setViewport({ width: 1280, height: 800 });

    for (const make of makes) {
        const makeFolder = make.replace(" ", "_");
        fs.mkdirSync(makeFolder, { recursive: true });

        const encodedMake = encodeURIComponent(make);
        const url = baseUrl.replace("{make}", encodedMake);

        await page.goto(url, { waitUntil: "networkidle2" });

        try {
            await page.waitForSelector("h3.text-black span.blue-heading", {
                timeout: 5000,
            });
            const totalResults = await page.$eval(
                "h3.text-black span.blue-heading",
                (el) => parseInt(el.innerText.replace(",", ""))
            );
            const lastPageNumber = Math.ceil(totalResults / 100);

            // Set items per page to 100
            await page.waitForSelector(".p-paginator-rpp-options", {
                timeout: 5000,
            });
            await page.click(".p-paginator-rpp-options");
            await new Promise((resolve) => setTimeout(resolve, 2000));
            await page.click("li[aria-label='100']");
            await new Promise((resolve) => setTimeout(resolve, 5000));

            const scrapePage = async (pageNumber) => {
                console.log(`Scraping page ${pageNumber}...`);

                await page.waitForSelector(
                    "tr.p-element.p-selectable-row.ng-star-inserted",
                    { timeout: 5000 }
                );

                const cars = await page.$$(
                    "tr.p-element.p-selectable-row.ng-star-inserted"
                );
                const carData = [];

                for (const car of cars) {
                    try {
                        const titleElement = await car.$(
                            "span.search_result_lot_detail"
                        );
                        const title = await titleElement.evaluate((el) =>
                            el.innerText.trim()
                        );

                        const linkElement = await car.$(
                            "a[aria-label='Lot Details']"
                        );
                        const link = await linkElement.evaluate(
                            (el) => el.href
                        );

                        const titleParts = title.split(" ");
                        const year = titleParts[0];
                        let model = title.replace(year, "").trim();

                        if (!title.includes(make)) {
                            if (make.includes(" ")) {
                                model = model
                                    .replace(make.replace(" ", "-").trim(), "")
                                    .trim();
                            } else if (make.includes("-")) {
                                model = model
                                    .replace(make.replace("-", " ").trim(), "")
                                    .trim();
                            }
                        } else {
                            model = model.replace(make, "").trim();
                        }

                        const bidElement = await car.$(
                            "span.search_result_amount_block.text-black.p-bold"
                        );
                        let currentBid = null;

                        if (bidElement) {
                            const amountElement = await bidElement.$(
                                "span.currencyAmount"
                            );
                            const currencyElement = await bidElement.$(
                                "span.currencyCode"
                            );

                            const amountText = await amountElement.evaluate(
                                (el) => el.innerText.trim()
                            );
                            const currency = await currencyElement.evaluate(
                                (el) => el.innerText.trim()
                            );

                            const amount = parseFloat(
                                amountText.replace(/[^\d.]/g, "")
                            );
                            currentBid = {
                                currency,
                                amount,
                            };
                        }

                        // Open car detail page
                        const newPage = await browser.newPage();
                        await newPage.goto(link, {
                            waitUntil: "networkidle2",
                        });
                        await new Promise((resolve) =>
                            setTimeout(resolve, 1000)
                        );

                        let details = null;

                        const lotNumber = await newPage
                            .$eval("#LotNumber", (el) => el.innerText.trim())
                            .catch(() => "N/A");
                        const imageUrl = await newPage
                            .$eval("#media-lot-image", (img) => img.src)
                            .catch(() => "N/A");
                        const thumbnails = await newPage.$$eval(
                            ".p-galleria-thumbnail-item img",
                            (images) => images.map((img) => img.src)
                        );
                        const vin = await newPage
                            .$eval(
                                "label[for='VIN:'] + div .lot-details-desc",
                                (el) => el.innerText.trim()
                            )
                            .catch(() => "N/A");
                        const primaryDamage = await newPage
                            .$eval(
                                "[data-uname='lotdetailPrimarydamagevalue']",
                                (el) => el.innerText.trim()
                            )
                            .catch(() => "N/A");
                        const color = await newPage
                            .$eval("[data-uname='lotdetailColorvalue']", (el) =>
                                el.innerText.trim()
                            )
                            .catch(() => "N/A");
                        const transmission = await newPage.evaluate(() => {
                            const labels = Array.from(
                                document.querySelectorAll("label")
                            );
                            for (let label of labels) {
                                if (
                                    label.innerText.trim() === "Transmission:"
                                ) {
                                    const span = label.nextElementSibling;
                                    if (
                                        span &&
                                        span.classList.contains(
                                            "lot-details-desc"
                                        )
                                    ) {
                                        return span.innerText.trim();
                                    }
                                }
                            }
                            return "N/A";
                        });
                        const drive = await newPage
                            .$eval("[data-uname='DriverValue']", (el) =>
                                el.innerText.trim()
                            )
                            .catch(() => "N/A");
                        const fuel = await newPage
                            .$eval("[data-uname='lotdetailFuelvalue']", (el) =>
                                el.innerText.trim()
                            )
                            .catch(() => "N/A");
                        const engineType = await newPage
                            .$eval("[data-uname='lotdetailEnginetype']", (el) =>
                                el.innerText.trim()
                            )
                            .catch(() => "N/A");

                        await newPage.close();

                        details = {
                            lotNumber,
                            vin,
                            primaryDamage,
                            transmission,
                            color,
                            drive,
                            fuel,
                            engineType,
                        };

                        const carInfo = {
                            title,
                            make,
                            year,
                            model,
                            link,
                            currentBid,
                            details,
                            imageUrl,
                            thumbnails,
                            uuid: uuidv4(),
                            createdAt: formattedYesterday,
                        };

                        console.log(carInfo);
                        carData.push(carInfo);
                    } catch (error) {
                        console.log("Error processing car:", error);
                    }
                }

                const filename = path.join(
                    makeFolder,
                    `car_data_page_${pageNumber}.json`
                );
                fs.writeFileSync(filename, JSON.stringify(carData, null, 4));
                console.log(`✅ Page ${pageNumber} data saved.`);
                return carData;
            };

            // Scrape first page
            await scrapePage(1);

            for (
                let currentPage = 1;
                currentPage < lastPageNumber;
                currentPage++
            ) {
                try {
                    const nextButton = await page.$("button.p-paginator-next");
                    const isDisabled = await page.evaluate(
                        (el) => el.classList.contains("p-disabled"),
                        nextButton
                    );
                    if (isDisabled) break;

                    await nextButton.click();
                    await new Promise((resolve) => setTimeout(resolve, 5000));
                    await scrapePage(currentPage + 1);
                } catch (error) {
                    console.log("Error navigating to the next page:", error);
                    break;
                }
            }

            console.log(`✅ All data saved in folder '${makeFolder}'`);
        } catch (error) {
            console.error("Error scraping:", error);
        }
    }

    await browser.close();
})();
